### PR TITLE
[Dataflow Streaming] Add a config to use multiple commit threads

### PR DESCRIPTION
**Pull request by arunpandianp**
Add a config to use multiple commit threads

### CODE REVIEW

**Comment by arunpandianp** [comment]
R: @scwhittle

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineDebugOptions.java [hunk 1]
```diff
@@ -260,6 +260,12 @@ public Dataflow create(PipelineOptions options) {
 
   void setstreamingSideInputCacheExpirationMillis(Integer value);
 
+  @Description("Number of commit threads used to commit items to streaming engine.")
+  @Default.Integer(1)
+  Integer getWindmillServiceCommitThreads();
```
Comment (arunpandianp) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineDebugOptions.java:265]: Where to add an integration test with a higher value?

Comment (scwhittle) [reply runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineDebugOptions.java:265]: You could add a test to StreamingDataflowWorkerTest.java
If you just want to cover that it functions, you could just set the option higher and run something like testBasic which expects 2000 commits.

If you want to more expliclitly test benefits of multiple commmit streams, you could modify the CommitWorkStream returned by FakeWindmillServer to support blocking on a large commit (simulating it being slow). Then you could have a test in StreamingDataflowWorkerTest where you have a large commit and smaller commits and verify that the smaller commits are received even while the large one is blocking.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java [hunk 2]
```diff
@@ -719,8 +726,10 @@ public void stop() {
       dispatchThread.join();
       // We need to interrupt the commitThread in case it is blocking on pulling
```
Comment (scwhittle) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:727]: nit: update comment for plural threads

**Review by scwhittle — APPROVED** [review]
See separate comment out of review about possible improved testing.

**Review by scwhittle — CHANGES_REQUESTED** [review]
see other comments, mistakenly approved last review

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineDebugOptions.java (hunk 1 above)
Comment (arunpandianp) [reply runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineDebugOptions.java:265]: Added testBasicWithMultipleCommitThreads to StreamingDataflowWorkerTest

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java (hunk 2 above)
Comment (arunpandianp) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java]: done

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java [hunk 3]
```diff
@@ -231,7 +231,7 @@ public class StreamingDataflowWorker {
   private final BoundedQueueExecutor workUnitExecutor;
   private final WindmillServerStub windmillServer;
   private final Thread dispatchThread;
-  private final Thread commitThread;
+  @VisibleForTesting final List<Thread> commitThreads;
```
Comment (scwhittle) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:234]: nit: how about ImmutableList and use builder below to create it.  Simpler to read code knowing it doesn't change size.

**Review by scwhittle — CHANGES_REQUESTED** [review]

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java (hunk 3 above)
Comment (arunpandianp) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java]: done

**Review by scwhittle — APPROVED** [review]

**Pull request merged**